### PR TITLE
GH-44360: [C#] Fix Flight DoExchange incompatibility with C++ implementation

### PR DESCRIPTION
**Pull request by adamreeve**
### Rationale for this change

See #44360

### What changes are included in this PR?

* Adds a new integration test to allow testing `do_exchange` between C++/Python and .NET.
* Updates the Flight stream reader to handle when a descriptor is sent in the first message without any schema.

### Are these changes tested?

* Yes, using the new integration test.

### Are there any user-facing changes?

No
* GitHub Issue: #44360

### CODE REVIEW

File: dev/archery/archery/integration/runner.py [hunk 1]
```diff
@@ -669,6 +669,12 @@ def append_tester(implementation, tester):
                          "RenewFlightEndpoint are working as expected."),
             skip_testers={"JS", "C#", "Rust"},
         ),
+        Scenario(
+            "do_exchange:echo",
+            description=("Test the do_exchange method by "
+                         "echoing data back to the client."),
+            skip_testers={"Go", "Java", "JS", "Rust"},
```
Comment (adamreeve) [inline dev/archery/archery/integration/runner.py:676]: I have a Java implementation working too but think that should be a separate PR.

**Comment by adamreeve** [comment]
I've also updated the status doc which had said DoExchange and Handshake were unsupported. It looks like Handshake support was added as part of #36079.

**Review by lidavidm — APPROVED** [review]
The non-C# bits look good to me.

**Review by CurtHagenlocher — APPROVED** [review]
